File: Source/Aladdin.CAPI/CAPI/TransportAgreement.cs
using System; 
using System.Diagnostics.CodeAnalysis; 

namespace Aladdin.CAPI
{
    ///////////////////////////////////////////////////////////////////////////
    // Алгоритм ассиметричного шифрования ключа на основе двух алгоритмов
    ///////////////////////////////////////////////////////////////////////////
    [SuppressMessage("Microsoft.Design", "CA1063:ImplementIDisposableCorrectly")]
    public class TransportAgreement : RefObject, ITransportAgreement 
    {
        // параметры алгоритма
        private ASN1.ISO.AlgorithmIdentifier parameters; 
    
        // создать алгоритм SSDH
        public static TransportAgreement CreateSSDH(Factory factory, 
            SecurityStore scope, ASN1.ISO.AlgorithmIdentifier parameters) 
        {
            // указать параметры шифрования ключа
            ASN1.ISO.AlgorithmIdentifier keyWrapParameters = 
                new ASN1.ISO.AlgorithmIdentifier(parameters.Parameters); 
        
            // создать алгоритм шифрования ключа
            using (IAlgorithm keyWrap = 
                factory.CreateAlgorithm<KeyWrap>(scope, keyWrapParameters))
            {
                // проверить поддержку алгоритма
                if (keyWrap == null) return null;  
            }
            // создать алгоритм согласования ключа
            using (IAlgorithm keyAgreement = 
                factory.CreateAlgorithm<IKeyAgreement>(scope, parameters))
            {
                // проверить поддержку алгоритма
                if (keyAgreement == null) return null;  
            }
            // создать алгоритм шифрования ключа
            return new TransportAgreement(parameters); 
        }
        // конструктор
        public TransportAgreement(ASN1.ISO.AlgorithmIdentifier parameters) 
        { 
            // сохранить переданные параметры
            this.parameters = parameters; 
        }
        // действия стороны-отправителя
        public virtual TransportAgreementData Wrap(IPrivateKey privateKey, 
            IPublicKey publicKey, IPublicKey[] recipientPublicKeys, IRand rand, ISecretKey key) 
        {
            // выделить буфер требуемого размера
            byte[][] encryptedKeys = new byte[recipientPublicKeys.Length][]; 
        
            // создать алгоритм согласования ключа
            using (KeyAgreement keyAgreement = CreateKeyAgreementAlgorithm(
                privateKey.Factory, privateKey.Scope, parameters))
            {    
                // сгенерировать случайные данные
                byte[] random = keyAgreement.Generate(publicKey.Parameters, rand); 

                // создать алгоритм шифрования ключа
                using (KeyWrap keyWrap = CreateKeyWrapAlgorithm(
                    privateKey.Factory, privateKey.Scope, parameters, random)) 
                {
                    // определить допустимые размеры ключей
                    int[] keySizes = keyWrap.KeySizes; int keySize = -1; 
        
                    // указать рекомендуемый размер ключа
                    if (keySizes != null && keySizes.Length == 1) keySize = keySizes[0]; 
        
                    // для всех получателей
                    for (int i = 0; i < recipientPublicKeys.Length; i++)
                    { 
                        // вычислить ключ шифрования ключа шифрования данных
                        using (ISecretKey KEK = keyAgreement.DeriveKey(privateKey, 
                            recipientPublicKeys[i], random, keyWrap.KeyFactory, keySize)) 
                        {
                            // проверить допустимость размера ключа
                            if (!KeySizes.Contains(keySizes, KEK.Length)) 
                            {
                                // выбросить исключение
                                throw new InvalidOperationException();
                            }
                            // зашифровать ключ
                            encryptedKeys[i] = EncodeEncryptedKey(keyWrap.Wrap(rand, KEK, key)); 
                        }
                    }
                }
                // вернуть зашифрованные ключи
                return new TransportAgreementData(publicKey, random, encryptedKeys); 
            }
        }
        // действия стороны-получателя
        public virtual ISecretKey Unwrap(IPrivateKey privateKey, IPublicKey publicKey, 
            byte[] random, byte[] encryptedKey, SecretKeyFactory keyFactory) 
        {
            // создать алгоритм согласования ключа
            using (KeyAgreement keyAgreement = CreateKeyAgreementAlgorithm(
                privateKey.Factory, privateKey.Scope, parameters))
            {    
                // создать алгоритм шифрования ключа
                using (KeyWrap keyWrap = CreateKeyWrapAlgorithm(
                    privateKey.Factory, privateKey.Scope, parameters, random)) 
                {
                    // определить допустимые размеры ключей
                    int[] keySizes = keyWrap.KeySizes; int keySize = -1;  
        
                    // указать рекомендуемый размер ключа
                    if (keySizes != null && keySizes.Length == 1) keySize = keySizes[0]; 
        
                    // вычислить ключ шифрования ключа шифрования данных
                    using (ISecretKey KEK = keyAgreement.DeriveKey(privateKey, 
                        publicKey, random, keyWrap.KeyFactory, keySize)) 
                    {
                        // проверить допустимость размера ключа
                        if (!KeySizes.Contains(keySizes, KEK.Length)) 
                        {
                            // выбросить исключение
                            throw new InvalidOperationException();
                        }
                        // расшифровать ключ
                        return keyWrap.Unwrap(KEK, DecodeEncryptedKey(encryptedKey), keyFactory); 
                    }
                }
            }
        }
        // закодировать/раскодировать зашифрованный ключ
        protected virtual byte[] EncodeEncryptedKey(byte[] encryptedKey) { return encryptedKey; }    
        protected virtual byte[] DecodeEncryptedKey(byte[] encryptedKey) { return encryptedKey; }    
    
        // получить алгоритм согласования ключа
        protected virtual KeyAgreement CreateKeyAgreementAlgorithm(
            Factory factory, SecurityStore scope, ASN1.ISO.AlgorithmIdentifier parameters)
        {
            // создать алгоритм согласования ключа
            KeyAgreement keyAgreement = (KeyAgreement)factory.CreateAlgorithm<IKeyAgreement>(scope, parameters); 
            
            // проверить наличие алгоритма
            if (keyAgreement == null) throw new NotSupportedException(); return keyAgreement; 
        }
        // получить алгоритм шифрования ключа
        protected virtual KeyWrap CreateKeyWrapAlgorithm(Factory factory, 
            SecurityStore scope, ASN1.ISO.AlgorithmIdentifier parameters, byte[] random)
        {
            // указать параметры шифрования ключа
            ASN1.ISO.AlgorithmIdentifier keyWrapParameters = 
                new ASN1.ISO.AlgorithmIdentifier(parameters.Parameters); 
        
            // создать алгоритм шифрования ключа
            KeyWrap keyWrap = factory.CreateAlgorithm<KeyWrap>(scope, keyWrapParameters); 
        
            // проверить поддержку алгоритма
            if (keyWrap == null) throw new NotSupportedException(); return keyWrap; 
        }
    }
}

File: Source/Aladdin.CAPI.Java/src/aladdin/capi/TransportAgreement.java
package aladdin.capi;
import aladdin.*; 
import aladdin.asn1.iso.*;
import aladdin.capi.keyx.*; 
import java.security.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////
// Алгоритм ассиметричного шифрования ключа на основе двух алгоритмов
///////////////////////////////////////////////////////////////////////////
public class TransportAgreement extends RefObject implements ITransportAgreement
{
    // параметры алгоритма
    private final AlgorithmIdentifier parameters; 
    
    // создать алгоритм SSDH
    public static TransportAgreement createSSDH(Factory factory, 
        SecurityStore scope, AlgorithmIdentifier parameters) throws IOException
    {
        // указать параметры шифрования ключа
        AlgorithmIdentifier keyWrapParameters = 
            new AlgorithmIdentifier(parameters.parameters()); 
        
        // создать алгоритм шифрования ключа
        try (IAlgorithm keyWrap = factory.createAlgorithm(
            scope, keyWrapParameters, KeyWrap.class))
        {
            // проверить поддержку алгоритма
            if (keyWrap == null) return null;  
        }
        // создать алгоритм согласования ключа
        try (IAlgorithm keyAgreement = factory.createAlgorithm(
            scope, parameters, IKeyAgreement.class))
        {
            // проверить поддержку алгоритма
            if (keyAgreement == null) return null;  
        }
        // создать алгоритм шифрования ключа
        return new TransportAgreement(parameters); 
    }
    // конструктор
    public TransportAgreement(AlgorithmIdentifier parameters) 
    { 
        // сохранить переданные параметры
        this.parameters = parameters; 
    }
    // действия стороны-отправителя
    @Override public TransportAgreementData wrap(IPrivateKey privateKey, 
        IPublicKey publicKey, IPublicKey[] recipientPublicKeys, IRand rand, ISecretKey key) 
        throws IOException, InvalidKeyException
    {
        // выделить буфер требуемого размера
        byte[][] encryptedKeys = new byte[recipientPublicKeys.length][]; 
        
        // создать алгоритм согласования ключа
        try (KeyAgreement keyAgreement = createKeyAgreementAlgorithm(
            privateKey.factory(), privateKey.scope(), parameters))
        {    
            // сгенерировать случайные данные
            byte[] random = keyAgreement.generate(publicKey.parameters(), rand); 

            // создать алгоритм шифрования ключа
            try (KeyWrap keyWrap = createKeyWrapAlgorithm(
                privateKey.factory(), privateKey.scope(), parameters, random)) 
            {
                // определить допустимые размеры ключей
                int[] keySizes = keyWrap.keySizes(); int keySize = -1; 
        
                // указать рекомендуемый размер ключа
                if (keySizes != null && keySizes.length == 1) keySize = keySizes[0]; 
        
                // для всех получателей
                for (int i = 0; i < recipientPublicKeys.length; i++)
                { 
                    // вычислить ключ шифрования ключа шифрования данных
                    try (ISecretKey KEK = keyAgreement.deriveKey(privateKey, 
                        recipientPublicKeys[i], random, keyWrap.keyFactory(), keySize)) 
                    {
                        // проверить допустимость размера ключа
                        if (!KeySizes.contains(keySizes, KEK.length())) 
                        {
                            // выбросить исключение
                            throw new IllegalStateException();
                        }
                        // зашифровать ключ
                        encryptedKeys[i] = encodeEncryptedKey(keyWrap.wrap(rand, KEK, key)); 
                    }
                }
            }
            // вернуть зашифрованные ключи
            return new TransportAgreementData(publicKey, random, encryptedKeys); 
        }
    }
    // действия стороны-получателя
    @Override public ISecretKey unwrap(IPrivateKey privateKey, IPublicKey publicKey, 
        byte[] random, byte[] encryptedKey, SecretKeyFactory keyFactory) throws IOException
    {
        // создать алгоритм согласования ключа
        try (KeyAgreement keyAgreement = createKeyAgreementAlgorithm(
            privateKey.factory(), privateKey.scope(), parameters))
        {    
            // создать алгоритм шифрования ключа
            try (KeyWrap keyWrap = createKeyWrapAlgorithm(
                privateKey.factory(), privateKey.scope(), parameters, random)) 
            {
                // определить допустимые размеры ключей
                int[] keySizes = keyWrap.keySizes(); int keySize = -1;  
        
                // указать рекомендуемый размер ключа
                if (keySizes != null && keySizes.length == 1) keySize = keySizes[0]; 
        
                // вычислить ключ шифрования ключа шифрования данных
                try (ISecretKey KEK = keyAgreement.deriveKey(privateKey, 
                    publicKey, random, keyWrap.keyFactory(), keySize)) 
                {
                    // проверить допустимость размера ключа
                    if (!KeySizes.contains(keySizes, KEK.length())) 
                    {
                        // выбросить исключение
                        throw new IllegalStateException();
                    }
                    // расшифровать ключ
                    return keyWrap.unwrap(KEK, decodeEncryptedKey(encryptedKey), keyFactory); 
                }
                // обработать неожидаемую ошибку
                catch (InvalidKeyException e) { throw new RuntimeException(e); }
            }
        }
    }
    // закодировать/раскодировать зашифрованный ключ
    protected byte[] encodeEncryptedKey(byte[] encryptedKey) throws IOException { return encryptedKey; }    
    protected byte[] decodeEncryptedKey(byte[] encryptedKey) throws IOException { return encryptedKey; }    
    
    // получить алгоритм согласования ключа
    protected KeyAgreement createKeyAgreementAlgorithm(
        Factory factory, SecurityStore scope, AlgorithmIdentifier parameters) throws IOException
    {
        // создать алгоритм согласования ключа
        KeyAgreement keyAgreement = (KeyAgreement)
            factory.createAlgorithm(scope, parameters, IKeyAgreement.class); 
            
        // проверить наличие алгоритма
        if (keyAgreement == null) throw new UnsupportedOperationException(); return keyAgreement; 
    }
    // получить алгоритм шифрования ключа
    protected KeyWrap createKeyWrapAlgorithm(Factory factory, 
        SecurityStore scope, AlgorithmIdentifier parameters, byte[] random) throws IOException
    {
        // указать параметры шифрования ключа
        AlgorithmIdentifier keyWrapParameters = new AlgorithmIdentifier(parameters.parameters()); 
        
        // создать алгоритм шифрования ключа
        KeyWrap keyWrap = (KeyWrap)factory.createAlgorithm(scope, keyWrapParameters, KeyWrap.class); 
        
        // проверить поддержку алгоритма
        if (keyWrap == null) throw new UnsupportedOperationException(); return keyWrap; 
    }
}
